zookeeper-metrics-providers/zookeeper-prometheus-metrics/src/main/java/org/apache/zookeeper/metrics/prometheus/PrometheusMetricsProvider.java

```diff
@@ -0,0 +1,295 @@
+            } finally {
+                server = null;
+            }
+        }
+    }
+
+    @Override
+    public void dump(BiConsumer<String, Object> sink) {
+        Enumeration<Collector.MetricFamilySamples> samplesFamilies = collectorRegistry.metricFamilySamples();
+        while (samplesFamilies.hasMoreElements()) {
+            Collector.MetricFamilySamples samples = samplesFamilies.nextElement();
+            samples.samples.forEach(sample -> {
+                StringBuilder keyBuilder = new StringBuilder();
+                keyBuilder.append(sample.name);
+                if (sample.labelNames.size() > 0) {
+                    keyBuilder.append('{');
+                    for (int i = 0; i < sample.labelNames.size(); ++i) {
+                        keyBuilder.append(sample.labelNames.get(i));
+                        keyBuilder.append("=\"");
+                        keyBuilder.append(sample.labelValues.get(i));
+                        keyBuilder.append("\",");
+                    }
+                    keyBuilder.append('}');
+                }
+                sink.accept(keyBuilder.toString(), sample.value);
+            });
+        }
+    }
+
+    @Override
+    public void resetAllValues() {
+        // not supported on Prometheus
+    }
+
+    private class Context implements MetricsContext {
+
+        private final ConcurrentMap<String, Gauge> gauges = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, PrometheusCounter> counters = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, PrometheusSummary> basicSummaries = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, PrometheusSummary> summaries = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, PrometheusLabelledSummary> basicSummarySets = new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, PrometheusLabelledSummary> summarySets = new ConcurrentHashMap<>();
+
+        @Override
+        public MetricsContext getContext(String name) {
+            // no hierarchy yet
+            return this;
+        }
+
+        @Override
+        public Counter getCounter(String name) {
+            return counters.computeIfAbsent(name, (n) -> {
+                return new PrometheusCounter(n);
+            });
+        }
+
+        @Override
+        public void registerGauge(String name, Gauge gauge) {
+            if (gauge == null) {
+                // null means 'unregister'
+                gauges.remove(name);
+            } else {
+                gauges.put(name, gauge);
+            }
+        }
+
+        @Override
+        public Summary getSummary(String name, DetailLevel detailLevel) {
+            if (detailLevel == DetailLevel.BASIC) {
+                return basicSummaries.computeIfAbsent(name, (n) -> {
+                    if (summaries.containsKey(n)) {
+                        throw new IllegalArgumentException("Already registered a non basic summary as " + n);
```

Comment:
The registry will handle this for you

Reply:
no, check basicSummaries vs summaries

Reply:
The client library will throw an exception if you try to register two Summary metrics with the same name.

Reply:
yes but here I want to tell that the caller is asking for the wrong DetailLevel, it is asking for DetailLevel.BASIC but that summary has already been asked as ADVANCED